crates/net/network/src/transactions.rs

```diff
@@ -1016,6 +1005,95 @@ struct Peer {
     client_version: Arc<String>,
 }
 
+// === TransactionFetcher ===
+/// * active requests - especially inflight hashes, so that whenever we receive new missing hashes
+///   we can check first if they're already being requested.
+/// * missing hashes and peers that send us these - so we can possibly re-request them
+
+#[derive(Debug, Default)]
+struct TransactionFetcher {
+    /// All currently active requests for pooled transactions.
+    inflight_requests: FuturesUnordered<GetPooledTxRequestFut>,
+    /// Set that tracks all hashes that are currently being fetched.
+    inflight_hash_to_fallback_peers: HashMap<TxHash, Vec<PeerId>>,
+}
+
+impl TransactionFetcher {
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> (PeerId, Option<Vec<PooledTransactionsElement>>, Option<RequestError>) {
+        let mut final_peer_id: Option<PeerId> = None;
+        let mut final_txs: Option<Vec<PooledTransactionsElement>> = None;
+        while let Poll::Ready(Some(GetPooledTxResponse { peer_id, result })) =
+            self.inflight_requests.poll_next_unpin(cx)
+        {
+            final_peer_id = Some(peer_id);
+
+            match result {
+                Ok(Ok(txs)) => {
+                    self.inflight_requests.clear();
+                    for tx in txs.0.iter() {
+                        self.inflight_hash_to_fallback_peers.remove(tx.hash());
+                    }
+                    final_txs = Some(txs.0);
+                }
+                Ok(Err(req_err)) => return (peer_id, None, Some(req_err)),
+                Err(_) => {
+                    // request channel closed/dropped
+                    return (peer_id, None, Some(RequestError::ChannelClosed))
+                }
+            }
+        }
+        if let Some(peer_id) = final_peer_id {
+            (peer_id, final_txs, None)
+        } else {
+            (Default::default(), None, Some(RequestError::BadResponse))
```

Comment:
this is only here because `test_handle_incoming_transactions` was failing. you might know a more elegant solution to that problem